### PR TITLE
Support nested user chrootenv environments

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change

Previously we couldn't run chrootenv-wrapped applications inside a chrootenv, we can now. My primary motivation was Lutris, although this is also helpful for development environments.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Infinisil** [comment]
Ah darn, unfortunately https://github.com/NixOS/nixpkgs/pull/62091 introduced a conflict. It seems that that PR isn't needed anymore after this one, but I'm not so sure on that.

File: pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c [hunk 1]
```diff
@@ -18,17 +18,34 @@
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <sys/wait.h>
+#include <sys/syscall.h>
 
-const gchar *bind_blacklist[] = {"bin", "etc", "host", "usr", "lib", "lib64", "lib32", "sbin", NULL};
+const gchar *bind_blacklist[] = {"bin", "etc", "host", "real-host", "usr", "lib", "lib64", "lib32", "sbin", NULL};
+
+int pivot_root(const char *new_root, const char *put_old) {
```
Comment (yegortimoshenko) [inline pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c]: Weird that glibc doesn't provide a wrapper for this syscall.

**Review by lukateras — APPROVED** [review]
LGTM! This is a huge improvement over current design, both on cleanup (tmpfs) and nesting fronts. I don't see any memory allocation issues.

**Comment by yegortimoshenko** [comment]
Also see https://github.com/NixOS/nixpkgs/pull/55973, which replaces chrootenv entirely with bubblewrap.

**Comment by abbradar** [comment]
Yup, wanted to review this one later as maybe we should just switch.

re allocations: learnt about `g_autofree` and related GCC attributes while working in this, was pleasantly surprised. 

On May 28, 2019 10:54:20 AM GMT+03:00, Yegor Timoshenko <notifications@github.com> wrote:
>Also see https://github.com/NixOS/nixpkgs/pull/55973, which replaces
>chrootenv entirely with bubblewrap.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/61740#issuecomment-496406308

-- 
Nikolay.

**Comment by Laaas** [comment]
@Infinisil mine is completely unneeded after this, since this PR doesn't do any nftw or ftw calls seemingly, which are buggy in glibc.

File: pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c [hunk 2]
```diff
@@ -18,17 +18,34 @@
 #include <sys/stat.h>
```
Comment (Laaas) [inline pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c]: You can remove the `#include <ftw.h>` above this (can't review that line unfortunately)

**Comment by abbradar** [comment]
Let's merge now; hopefully we can replace all this code with bwrap in the end ;)

**Pull request merged**